src/modules/tools/iremitter/Iremitter.cpp

```diff
@@ -0,0 +1,294 @@
+*/
+
+#include "Kernel.h"
+#include "Iremitter.h"
+#include "checksumm.h"
+#include "Config.h"
+#include "ConfigValue.h"
+#include "Gcode.h"
+#include "Conveyor.h"
+#include "StreamOutputPool.h"
+
+// config names
+#define iremitter_checksum CHECKSUM("iremitter")
+#define enable_checksum    CHECKSUM("enable")
+#define mcode_checksum     CHECKSUM("mcode")
+#define pin_checksum       CHECKSUM("pin")
+#define camera_checksum    CHECKSUM("camera")
+
+Iremitter::Iremitter() {}
+
+void Iremitter::on_module_loaded()
+{
+	// if the module is disabled -> do nothing
+	if(! THEKERNEL->config->value(iremitter_checksum, enable_checksum)->by_default(false)->as_bool()) {
+		// as this module is not needed free up the resource
+		delete this;
+		return;
+	}
+
+	// settings
+	this->on_config_reload(this);
+
+	// events
+	this->register_for_event(ON_GCODE_RECEIVED);
+	this->register_for_event(ON_GCODE_EXECUTE);
+}
+
+void Iremitter::on_config_reload(void *argument)
+{
+	this->mcode = THEKERNEL->config->value(iremitter_checksum, mcode_checksum)->by_default(0)->as_int();
+	this->pin.from_string(THEKERNEL->config->value(iremitter_checksum, pin_checksum)->by_default("nc")->as_string())->as_output();
+	string camera = THEKERNEL->config->value(iremitter_checksum, camera_checksum)->by_default("nikon")->as_string();
+	
+	if      (camera == "canon")        this->camera = CANON;
+	else if (camera == "canonwldc100") this->camera = CANONWLDC100;
+	else if (camera == "minolta")      this->camera = MINOLTA;
+	else if (camera == "nikon")        this->camera = NIKON;
+	else if (camera == "olympus")      this->camera = OLYMPUS;
+	else if (camera == "pentax")       this->camera = PENTAX;
+	else if (camera == "sony")         this->camera = SONY;
+	else 
+	{
+		THEKERNEL->streams->printf("iremitter: camera not found, 'Nikon' set by default.\n");
+		this->camera = NIKON;
+	}
+}
+
+bool Iremitter::is_executable(Gcode *gcode)
+{
+	// return true if all condition for execution is satisfied...
+	return this->pin.connected() and this->mcode > 0 and gcode->has_m and gcode->m == this->mcode;
+}
+
+void Iremitter::on_gcode_received(void *argument)
+{
+	// received gcode
+	Gcode *gcode = static_cast<Gcode *>(argument);
+
+	// if executable, append to the queue
+	if (this->is_executable(gcode))
+		THEKERNEL->conveyor->append_gcode(gcode);
+}
+
+void Iremitter::on_gcode_execute(void *argument)
+{
+	// executed gcode
+	Gcode *gcode = static_cast<Gcode *>(argument);
+
+	// exit if it is not executable
+	if (! this->is_executable(gcode)) 
+		return;
+
+	// take a shot
+	this->trigger();
+}
+
+/*
+	original code from Sébastian Setz at :
+	http://sebastian.setz.name/arduino/my-libraries/multi-camera-iremitter-control/
+*/
+
+// pwm at freqency for x microseconds
+void Iremitter::high_us(unsigned int us, int freq)
+{
+	uint32_t start = us_ticker_read();
+	int pause = (1000 / freq / 2) - 4;
+
+	while((us_ticker_read() - start) < us)
+	{
```

Comment:
I think you need to use a timer for this... busy waits are not very friendly to the rest of the code.